lib/services/role.go

```diff
@@ -118,41 +139,390 @@ type Access interface {
-	// SetForwardAgent sets forward agent property
-	SetForwardAgent(forwardAgent bool)
-	// CheckAndSetDefaults checks and set default values for missing fields.
+	// CheckAndSetDefaults checks and set default values for any missing fields.
 	CheckAndSetDefaults() error
-	// Equals returns true if roles are equal
+	// Equals returns true if the roles are equal.
 	Equals(other Role) bool
+
+	// GetOptions gets role options.
+	GetOptions() RoleOptions
+	// SetOptions sets role options
+	SetOptions(opt RoleOptions)
+
+	// GetLogins gets *nix system logins for allow or deny condition.
+	GetLogins(RoleConditionType) []string
+	// SetLogins sets *nix system logins for allow or deny condition.
+	SetLogins(RoleConditionType, []string)
+
+	// GetNamespaces gets a list of namespaces this role is allowed or denied access to.
+	GetNamespaces(RoleConditionType) []string
+	// GetNamespaces sets a list of namespaces this role is allowed or denied access to.
+	SetNamespaces(RoleConditionType, []string)
+
+	// GetNodeLabels gets the map of node labels this role is allowed or denied access to.
+	GetNodeLabels(RoleConditionType) map[string]string
+	// SetNodeLabels sets the map of node labels this role is allowed or denied access to.
+	SetNodeLabels(RoleConditionType, map[string]string)
+
+	// GetRules gets all allow or deny rules.
+	GetRules(rct RoleConditionType) map[string][]string
+	// SetRules sets an allow or deny rule.
+	SetRules(rct RoleConditionType, rrs map[string][]string)
+}
+
+// RoleV3 represents role resource specification
+type RoleV3 struct {
+	// Kind is the type of resource.
+	Kind string `json:"kind"`
+	// Version is the resource version.
+	Version string `json:"version"`
+	// Metadata is resource metadata.
+	Metadata Metadata `json:"metadata"`
+	// Spec contains resource specification.
+	Spec RoleSpecV3 `json:"spec"`
+}
+
+// Equals returns true if roles are equal.
+func (r *RoleV3) Equals(other Role) bool {
+	return true
+}
+
+// SetExpiry sets expiry time for the object.
+func (r *RoleV3) SetExpiry(expires time.Time) {
+	r.Metadata.SetExpiry(expires)
+}
+
+// Expiry returns the expiry time for the object.
+func (r *RoleV3) Expiry() time.Time {
+	return r.Metadata.Expiry()
+}
+
+// SetTTL sets TTL header using realtime clock.
+func (r *RoleV3) SetTTL(clock clockwork.Clock, ttl time.Duration) {
+	r.Metadata.SetTTL(clock, ttl)
+}
+
+// SetName sets the role name and is a shortcut for SetMetadata().Name.
+func (r *RoleV3) SetName(s string) {
+	r.Metadata.Name = s
+}
+
+// GetName gets the role name and is a shortcut for GetMetadata().Name.
+func (r *RoleV3) GetName() string {
+	return r.Metadata.Name
+}
+
+// GetMetadata returns role metadata.
+func (r *RoleV3) GetMetadata() Metadata {
+	return r.Metadata
+}
+
+// GetOptions gets role options.
+func (r *RoleV3) GetOptions() RoleOptions {
+	return r.Spec.Options
+}
+
+// SetOptions sets role options.
+func (r *RoleV3) SetOptions(ro RoleOptions) {
+	// TODO(russjones): deep copy here?
```

Comment:
I would not do deep copy here, instead I would add shortcut to simplify updating options in place, so you would not need to expose underlying dictionary in the first place